libs/core/langchain_core/documents/base.py

```diff
@@ -1,47 +1,269 @@
 
-            document = Document(
-                page_content="Hello, world!",
-                metadata={"source": "https://example.com"}
-            )
+    All such objects are expected to allow specifying metadata and an
+    optional identifier.
     """
 
     # The ID field is optional at the moment.
     # It will likely become required in a future major release after
     # it has been adopted by enough vectorstore implementations.
     id: Optional[str] = None
     """An optional identifier for the document.
-    
+
     Ideally this should be unique across the document collection and formatted 
     as a UUID, but this will not be enforced.
     """
 
-    page_content: str
-    """String text."""
     metadata: dict = Field(default_factory=dict)
-    """Arbitrary metadata about the page content (e.g., source, relationships to other
-        documents, etc.).
+    """Arbitrary metadata associated with the content."""
+
+
+class Blob(BaseContent):
+    """Blob represents raw data by either reference or value.
+
+    Provides an interface to materialize the blob in different representations, and
+    help to decouple the development of data loaders from the downstream parsing of
+    the raw data.
+
+    Inspired by: https://developer.mozilla.org/en-US/docs/Web/API/Blob
+
+    Example: Initialize a blob from in-memory data
+
+        .. code-block:: python
+
+            from langchain_core.documents import Blob
+
+            blob = Blob.from_data("Hello, world!")
+
+            # Read the blob as a string
+            print(blob.as_string())
+
+            # Read the blob as bytes
+            print(blob.as_bytes())
+
+            # Read the blob as a byte stream
+            with blob.as_bytes_io() as f:
+                print(f.read())
+
+    Example: Load from memory and specify mime-type and metadata
+
+        .. code-block:: python
+
+            from langchain_core.documents import Blob
+
+            blob = Blob.from_data(
+                data="Hello, world!",
+                mime_type="text/plain",
+                metadata={"source": "https://example.com"}
+            )
+
+    Example: Load the blob from a file
+
+        .. code-block:: python
+
+            from langchain_core.documents import Blob
+
+            blob = Blob.from_path("path/to/file.txt")
+
+            # Read the blob as a string
+            print(blob.as_string())
+
+            # Read the blob as bytes
+            print(blob.as_bytes())
+
+            # Read the blob as a byte stream
+            with blob.as_bytes_io() as f:
+                print(f.read())
     """
+
+    data: Union[bytes, str, None]
+    """Raw data associated with the blob."""
+    mimetype: Optional[str] = None
+    """MimeType not to be confused with a file extension."""
+    encoding: str = "utf-8"
+    """Encoding to use if decoding the bytes into a string.
+
+    Use utf-8 as default encoding, if decoding to string.
+    """
+    path: Optional[PathLike] = None
```

Comment:
Is this important/general enough to be top level or does it belong in metadata?

Reply:
It's used to allow specifying the Blob by reference rather than by value (i.e., a URL). We could've probably encoded this information using `data` + `encoding` . Right now need to keep it for backwards compatibility and we could revisit in the future.